ruby/phase-1.md

```diff
@@ -0,0 +1,616 @@
+
+```ruby
+class MyClass
+  def my_method
+    puts 'Hello World!'
+  end
+end
+
+class MyClass
+  def my_method
+    puts 'Hello Monkeys!'
+  end
+
+  def new_method
+    puts 'Hello New World!'
+  end
+end
+
+my = MyClass.new
+my.my_method
+# => Hello Monkeys!
+my.new_method
+# => Hello New World
+```
+
+### Concurrency
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
+executed at the same time (making it effectively single-threaded). The good news is that GIL does
+allow IO-bound instructions to be executed in parallel.
+
+While threads can offer a great performance boost in applications with a significant IO reliance. In
+order to effectively utilise a multi-core processor a Ruby application has to be executed using
+multiple processes.
+
+#### Fibers
+
+Ruby also has Fibers that allow light weight cooperative concurrency. These are code blocks that can
+be paused (from inside) and resumed (from outside). These are really useful for controlling the
+execution of workflow code. However unlike threads Fibers are never preempted and their scheduling
+has to be done manually by the code:
+
+```ruby
+fiber = Fiber.new do
+  i = 0
+  loop do
+    i += 1
+    puts i
+    Fiber.yield
+  end
+end
+
+fiber.resume
+# => 1
+fiber.resume
+# => 2
+fiber.resume
+# => 3
+```
+
+*NOTE: Each `fiber.resume` call blocks the main thread until a Fiber `yield`s back control.*
+
+#### Reactors
+
+Somewhat less popular, but still very viable are libraries that implement a reactor pattern for
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
```

Comment:
From the public API point of view SDK exposes only 2 components — Client and Worker. The Configuration tries to capture the shared configuration options between the two (and a gRPC client by proxy since both Client and Worker depend on it). I agree that there will be options specific to Client and Worker and these will be passed as keyword arguments on initialization (capturing the options specific to either Client or Worker).

Reply:
gRPC client is public as well. How are you going to make requests like listing workflows?

Reply:
I'd like to challenge this — I believe that gRPC client is only temporarily public until there are higher-level abstractions on top of all publicly accessible calls. The reasoning is simple — there are already some methods on the gRPC client today that are not expected to be called directly (like high-level Client methods or Worker methods) and the whole reason behind an SDK is to not force people to interact with the low-level API.

If this is NOT the case — why no let people use gRPC client directly without having to go via the high-level client? I would expect a 2-part SDK in that case, so you can choose between a high-level client and a low-level client. Having to use both is confusing (they have very different APIs).

Reply:
> If this is NOT the case — why no let people use gRPC client directly without having to go via the high-level client?

It's not the case sometimes and is the case others. We are intentionally not absolute here. We have higher level abstractions for starting and doing things with workflows and we intentionally do not provide abstractions for doing things like listing workflows, creating namespaces, etc. It's unreasonably to think we should provide high level abstractions for all calls since things are commonly added (Go did this and we regret it and have since exposed the raw gRPC client), and similarly unreasonable to expect them to navigate the complexities that the high level client does for things like fetching a result.